roles/openshift_health_checker/openshift_checks/sdn.py

```diff
@@ -0,0 +1,354 @@
+        """
+        self.register_file(path, self.read_command_output(command))
+
+    def read_command_output(self, command):
+        """Execute the provided command using the command module
+        and return its output.
+
+        If the command is a string, use a shell.  Otherwise, assume the command
+        is a list, join it with spaces, and execute it without shell.
+        """
+        uses_shell = False
+        if isinstance(command, six.string_types):
+            uses_shell = True
+        else:
+            command = ' '.join(command)
+
+        command_args = dict(_raw_params=command, _uses_shell=uses_shell)
+        # Use self._execute_module instead of self.execute_module because
+        # the latter sets self.changed.
+        result = self._execute_module('command', command_args)
+        if result.get('rc', 0) != 0 or result.get('failed'):
+            raise OpenShiftCheckException(
+                'RemoteCommandFailure',
+                'Failed to execute command on remote host: %s' % command)
+
+        return result['stdout'].encode('utf-8')
+
+    def check_master(self):
+        """Gather diagnostic information on a master and ensure it can connect
+        to kubelets."""
+        if self.want_full_results:
+            conf_base_path = self.get_var('openshift.common.config_base')
+            master_conf_path = os.path.join(conf_base_path, 'master',
+                                            'master-config.yaml')
+            self.register_file('master-config.yaml', None, master_conf_path)
+
+            self.save_component_container_logs('controllers', 'controllers')
+            self.save_component_container_logs('api', 'api')
+
+        nodes = self.get_resource('nodes')
+        pods = self.get_resource('pods')
+        services = self.get_resource('services')
+
+        if self.want_full_results:
+            self.register_file('nodes.json', nodes)
+            self.register_file('pods.json', pods)
+            self.register_file('services.json', services)
+            self.register_file('endpoints.json',
+                               self.get_resource('endpoints'))
+            self.register_file('routes.json', self.get_resource('routes'))
+            self.register_file('clusternetworks.json',
+                               self.get_resource('clusternetworks'))
+            self.register_file('hostsubnets.json',
+                               self.get_resource('hostsubnets'))
+            self.register_file('netnamespaces.json',
+                               self.get_resource('netnamespaces'))
+
+        if not nodes:
+            self.register_failure(
+                'No nodes appear to be defined according to the API.'
+            )
+
+        for node in nodes:
+            self.check_node_kubelet(node)
+
+    def save_component_container_logs(self, component, container):
+        """Save the first and last 500 lines of logs for the specified
+        component and container."""
```

Comment:
Not sure if first 500 lines will be helpful but prefer more than 500 lines from the last.
We dump a lot of data to the logs at level 5 and last 500 lines may not provide sufficient data to debug.

Reply:
First 500 lines show startup.  I'll increase to first and last 2000; is that adequate? We want to avoid transferring more than a few megabytes because it will be slurped into memory, base-64 encoded, and then transferred over the network.

Sidenote: Dan Winship said, "Definitely want to use time rather than number of lines, since we don't know what debug level they're using." (https://github.com/openshift/openshift-ansible/pull/7759#discussion_r179188089).  However, the container logs don't have dates, only HH:MM:SS, so we cannot reliably compute intervals.